certbot/plugins/script.py

```diff
@@ -0,0 +1,172 @@
+"""Standalone Authenticator."""
+import logging
+import os
+import sys
+
+import zope.interface
+
+from subprocess import Popen, PIPE
+
+from acme import challenges
+
+from certbot import errors
+from certbot import interfaces
+
+from certbot.plugins import common
+
+logger = logging.getLogger(__name__)
+
+
+CHALLENGES = ["http-01", "dns-01"]
+
+
+@zope.interface.implementer(interfaces.IAuthenticator)
+@zope.interface.provider(interfaces.IPluginFactory)
+class Authenticator(common.Plugin):
+    """Script authenticator
+
+    calls user defined script to perform authentication and
+    optionally cleanup.
+
+    """
+
+    description = "Authenticate using user provided script(s)"
+
+    def __init__(self, *args, **kwargs):
+        super(Authenticator, self).__init__(*args, **kwargs)
+        self.cleanup_script = None
+        self.auth_script = None
+        self.challenges = []
+
+    @classmethod
+    def add_parser_arguments(cls, add):
+        add("auth", default=None, required=False,
+            help="path to the authentication script")
+        add("cleanup", default=None, required=False,
+            help="path to the cleanup script")
+
+    @property
+    def supported_challenges(self):
+        """Challenges supported by this plugin."""
+        return self.challenges
+
+    def check_script_validity(self, script_path):
```

Comment:
Now using hooks.validate_hook, had to do a bit of refactoring to hooks to accomplish this.